### ISSUE TITLE
Reduction of Bandwidth/Traffic Consumption

### ISSUE THREAD

**Issue by srkunze** [irungentoo/toxcore#1251]
Is there something we can do about it?

https://github.com/subliun/Antox/issues/6

**Comment by ollieh** [comment]
Mobile apps aren't really supposed to be constantly connected to servers, but to use the push notification system the OS provides. I can't think of a way to do this without using a centralized server. The best situation I can think of is a centralized server just for sending the push notifications, and then the mobile app sends a message to the central push notification server authorizing the TCP relay it is connected to to send everything it receives for that user to the push notification server. It's not such an issue on Android, because you can have stuff working in the background, and we were getting half-decent battery life at one point so it is possible. You can't get anywhere near the battery efficiency of push notifications though, and for iOS it's a complete showstopper. On iOS you can't have things running in the background for more than a few seconds or maybe a minute, so a messaging app on iOS can't work in the background, and must use push messages if it wants to actually be useful. I think it also means user-written messages would have to go through the push system, not just friend requests and such. I may be wrong but it doesn't look promising. Can anyone see an easier way out?

**Comment by srkunze** [comment]
@ollieh I think what you said is really important. Not sure how push notifications work. If the app can configure the push server freely, it could be changed to respective peers regularly. Could this work?

I want to emphasize the fact that mobile will be (is already?) the **dominant** way people use chat/voice/video applications. A solution here is crucial (at least in my opinion). Most people will never (exaggerated of course) use two systems (Skype+Tox) because of the network effect (friends etc.).

**Comment by ollieh** [comment]
From the android docs:
>Send data from your server to users' Android-powered devices
>This could be a lightweight message telling your app there is new data to be fetched from the server or...

It seems like this approach might work. The TCP relay could hold the messages for a little, send a message to the push server to tell the user's client to query the TCP relay, and then pass the messages on when it does. That way you wouldn't necessarily need to authorize each TCP relay as the worst thing they could do would be constantly send push messages that tell you to query the network, to drain your battery. As it'd be a central server controlling it you could just have a timer on it. I wish there was a way to do this without a central server though.

I totally agree with you about mobile. If Tox is to succeed, it has to be fantastic on mobile. Everyone seems to have their sights set on beating Skype, but we can't beat Whatsapp yet. iOS is impossible at the moment, and I don't know about you, but most of my friends have iPhones, and Tox isn't very useful if your friends can't use it. I know most of us probably prefer using desktop applications, but if it's desktop only it'll never take off. If it just did mobile well it'd stand a much, much better chance, even if the desktop clients were shitty or non-existent. We need to solve push messages, and offline messages.

**Comment by tttom** [comment]
@ollieh, I second that. I'd say that the average internet user does not have a computer anymore, but almost certainly just a smartphone of some kind. Battery life and data usage cost of smartphones practically mandate the use of iOS push notifications or Google Cloud messaging. In principle one could sign up to one or more message services of choice, which would keep encrypted messages and send push notifications to their phone. Although I believe that the push notification service is free of charge, running a message service server to send the push notifications would not be entirely free. Thing is, I don't see any incentive in running such a message service unless it is for oneself or ones friends. Bittorrent works because uploaders get somehow rewarded with faster download speeds. Free e-mail providers get to peek into your messages to increase their ad revenue. But here, message service providers would just pay for the bandwidth of smartphone users. Hence, my honest question, is there anything that smartphone users can offer in return for good battery life and minimal data costs?

**Comment by srkunze** [comment]
@tttom You are right. Not sure if there could be a charity organisation which can collect some money and distribute it to node operators to cover their expenses.

However, we should give people a sensible option first on mobile before thinking about compensation.

**Comment by srkunze** [comment]
Adhering to the initial problem, @dubslow estimated here https://github.com/irungentoo/toxcore/issues/1254#issuecomment-74318898 that one package/message could be 1-2 KiB.

Unfortunately, the tablet traffic monitors tells us about 50 KiB per message. What is wrong here?

Is onion routing a problem?

**Comment by dubslow** [comment]
There are indeed more packets than just the message packets...

**Comment by srkunze** [comment]
25 fold more?

**Comment by Kolargol00** [comment]
I'm seeing Antox use a huge amount of data on wifi: in the range of 10Mb to 50MB per hour while I'm not using the app at all. Is there any number/log I can report to help you guys diagnose this?

**Comment by fcore117** [comment]
i hope there are more TCP optimizations, could be useful especially for 2G, 3G, 4G internet users and for very low bandwith ADSL users.

**Comment by dcharles525** [comment]
This has to be fixed, or Tox on the mobile platform will fail.

**Comment by LapFox** [comment]
Tox on desktop can also be under threat, for people with data caps on their broadband (it's more common than you think)

**Comment by fcore117** [comment]
LapFox: this is very common outside of city like farms and houses in woods, mobile broadband is only option for desktop. I myself even have helped some people to set up usb stick to get better signal. Most common monthly plan is 5-20GB and only very rich could get 100GB to infinite bandwidth.

**Comment by srkunze** [comment]
Is there anything the community can do here?

**Comment by Kolargol00** [comment]
Disabling UDP and installing a recent Antox build brought down data usage to much lower levels (< 1Mb / day). Unfortunately, I can't say which operation changed that since I did both together.

**Comment by srkunze** [comment]
What the hell?

**Comment by lovetox** [comment]
you should give this issue the highest priority

as some said before, tox can and never will be the future of online communication, if it is unusable on smartphones.

**Comment by grinapo** [comment]
Maybe it would help to track down the issues to generate stats about the various kinds of traffic tox uses and see what uses how much, and after it could be checked why and whether it can be shrunk or not. My guess (without looking) is that it may be an inherent design problem with the protocol and would go the same as bleep: unusable on mobile due to extreme high protocol traffic overhead.

**Comment by optimumtact** [comment]
This issue has become much more important, because the upcoming versions of android will require apps to properly sleep and only wake up on messages from google push notifications.

AFAIK IOS already does this

**Comment by timofonic** [comment]
I see too much discussion, that's great.

Are there some solution without sacrificing p2p? IPhone can be jailbreaked, right?

-- Technological illiteracy: The real issue --

<This pseudo-essay is based on personal experiences, without enough research. Probably there's already better essays on the covered concepts, but I didn't find them. 
- A study on this is very hard to do because it requires enough knowledge on mind sciences, society and technology.
* Knowledge segmentation and specialization makes it nearly impossible to do it properly. This is both an educational and biological issue.
* People are horrible at teamworkin

I always recommend iPhone/Android users to get more advanced and take control of their systems, but it's something difficult to change in very stubborn i-don't-like-computer-systems-but-use-them-alot-fix-this-asap people.

Really, it's shameful how people is getting even more technically illiterate than you may think:
- Malware is in all their systems most of their time and even don't care or are unable to learn how to do basic maintenance of their systems. They usually parasite on a "geek" and see us as mere tools.
- They are obsessively social, yet they only don't get enough curiosity to learn basic technology concepts.
* Women are usually faster at learning than most men if you remove stress and negativity to the topic, something very difficult to do but not impossible. I've seen most women in my environment to become not interested at all or just subjectively pragmatic interests such as social networks, casual games, messaging and information consumption. This subject affects me personally in too many life aspects, including sentimental ones.
* This is a social problem that it's really difficult to solve

**Comment by ameenross** [comment]
> They usually parasite on a "geek" and see us as mere tools.

QFT :laughing:
"Geeks" should probably just charge more!

**Comment by timofonic** [comment]
I agree!

We often underestimate our skills. We usually get parasited and exploited
most of the time, instead using our extra efforts for more interesting
things (like FOSS and learn new skills/knowledge).


Back to topic...

I think bw us very important. I would like to know why subliun considers
Opus better than Codec2. I reckon it's not good for general audio, but I
see it as the perfect sketch speech codec to avoid the usual bandwidth
problem in VoIP

**Comment by subliun** [comment]
There is no p2p solution to this problem. Antox will adopt GCM to send wakeup pings/messages sometime in the future, which should mostly remedy the bandwidth issue. I am sure improvements can be made in toxcore to improve bandwidth usage also.

**Comment by quininer** [comment]
@subliun GCM? You can describe the realization of the principle of it? It requires a special server?

**Comment by timofonic** [comment]
@subliun

Hello.

Thanks for your information. I'll uninstall Antox inmediatelly in the
version that adopts Google Cloud Messaging (
https://developers.google.com/cloud-messaging/ ). I already uninstalled all
third party information providers (including Google ones) and control al
flux of information by using AFWall+ installed with Fdroid.

I consider this a security issue and I'm against cloud computing services.
I'm going to migrate my email to my own email server for that reason too.

It seems popularity is a lot more important than security. I don't care
about popularity but security and dependence, so I'll not continue to use
this software if chooses the "popularity" over real security.

Kind regards.

**Comment by GrayHatter** [comment]
@subliun irssi has a nice plugin that encrypts the messages using a shared key so that neither the app, nor google can see what's in the message. I see no reason you couldn't just push the encrypted messaged to your app.

**Comment by subliun** [comment]
@GrayHatter everything going through GCM will be encrypted of course.

@timofonic The F-Droid build will likely never contain any GCM-related functionality. Unfortunately, due to restrictions imposed by Google in Android M, Antox is forced to adopt GCM or be left crippled and unable to receive phone calls, file transfers or messages in the background. Because of this, the Google Play build of Antox will be using GCM to deliver messages and updates in the background.

**Comment by tttom** [comment]
@subliun, as sending Google Cloud Messages requires each push server to have a Google API key, how would this be implemented for Antox? Ideally all tox nodes should be able to send GCM to antox users, though I don't see it happen that everybody will set up an account with Google, and later Apple etc. Would each Antox user have to run such a push server themselves or would there be a hard coded antox.net server perhaps?

Alternatively, with modifications to toxcore, one could designate one or more relays to be contacted through. Google Play users could register with a relay that is capable of reaching them by GCM and inform their contacts of this, i.e. an extended toxid. Relays would only see encrypted messages destined for that specific Antox user, hence one doesn't necessarily need to run their own service. The ability to specify multiple relays would enable high availability with low cost hardware, and perhaps even offline messaging. The configuration could remain as simple by setting up a default in the client. Users that don't like relays can specify no relay, and watch their battery drain... The designated relay principle could also be used for iOS as well as any open source alternatives for GCM.

**Comment by subliun** [comment]
>Would each Antox user have to run such a push server themselves or would there be a hard coded antox.net server perhaps?
This is accurate. My current plan is for some extension in toxcore to allow telling your friend somehow that you want your messages relayed over GCM. Then when you go offline messages will be sent to an offline messaging server/TCP relay server/to be determined, which will forward to an antox.me server set up to handle rate-limiting and abuse prevention. This server will have the GCM api key and will send the messages/wakeup updates through GCM which will be received in the background by the recipient.

I'm open to future extensions which allow more customisation or specification of custom forwarding servers, though I'd like to get the basics worked out first.

**Comment by quininer** [comment]
@subliun 
Perhaps you can write a similar relay server ZNC, allowing users to build their own.
Special node is not a good idea. According to this idea, the future may also have dedicated node iOS, WP, Blackberry OS.

**Comment by fcore117** [comment]
so Antox will not work with TCP mode too to send files and so on like toxcore should?

I am user who has no Google services installed, pure slim barebone.

**Comment by subliun** [comment]
@fcore117 it will, but not in the background. Please read https://commonsware.com/blog/2015/06/03/random-musing-m-developer-preview-ugly-part-one.html as it covers the issue very well.

**Comment by tttom** [comment]
@subliun, granted, multiple relays would perhaps be a bit more complex. I totally agree to try to keep things as simple as practical. Since you are already considering extending toxcore to tell friends they can relay messages over GCM, why not just tell them they can relay over a given toxid? That way the node at the specified toxid can implement the push notification system of choice. Antox could default to specify a toxid where antox.me sits, while future clients can do as they please.
Am I missing something, or isn't this less complex than something GCM/antox.im specific?

**Comment by xavierle** [comment]
@subliun 
maybe we can implement ourbown push server and not gcm

one example
https://github.com/WhisperSystems/PushServer

some others are existing

**Comment by fcore117** [comment]
would it not just be possible to make something like "idle TCP" mode that is great when client is in idle mode and when chat begins it tries to make direct connection or relayed depending of connection? 

Devs what are your thoughts currently? i really hope it could be internal feature without any servers.

**Comment by LuccoJ** [comment]
Please don't use GCM. Make, and keep, this app fully free and open, with no proprietary dependencies. Doing otherwise would defeat any attempt to have a really secure (and transparent) messaging system. I'm sure if enough developers refuse to cave in to GCM to sidestep Doze and App Standby, something will be done at Google or elsewhere to fix the matter.

**Comment by xavierle** [comment]
maybe mozilla notifications (simplepush or webpush api ) may be an optional
function to save bandwidth

until a better way is found

at least i would prefer a mozilla api or protocol than a Google one

**Comment by LuccoJ** [comment]
@xavierle I don't know about those, but I suspect only Google (or at least something installed in the Android firmware itself) can provide push notifications that will override Doze and App Standby. Other libraries will be subjected to the same limitations that normal apps have under Android 6.0.

**Comment by GrayHatter** [comment]
The problem isn't data, that's easy. The issue is cpu cycles, and there by battery usage. We need a way to maintain a connection to [thing] that can sleep aggressively, wake up the system when it needs to, and is simple to implement and maintain.

The reason GCM fits that bill is because it has a good team behind it (Google), an easy API, and is built DEEP into the Android OS.

Also GCM is FOSS btw... https://github.com/google/gcm/blob/master/LICENSE so maybe avoid the FUD?

**Comment by LuccoJ** [comment]
@GrayHatter "This project contains client libraries and samples to help developers interface with and explore the Google Cloud Messaging APIs." is straight from the thing you linked. GCM itself (both the server and the API on Android) is most definitely NOT free software, as anyone behind F-Droid will tell you if you ask them.

It is not "built into" Android at all, as in fact, it's not part of AOSP ("Android"), but only of the Google Play Services. This is, again, something anyone from F-Droid will confirm (try asking in #fdroid on freenode if you like).

**Comment by xavierle** [comment]
like every google service advantage is easy to use.
but gcm is clearly not free.

and i like im project without gcm like conversations, freelab messenger and
antox.

including gcm would be a mistake in my opinion.

if a push service is necessary to lower the battery consumption.
maybe we can use our own push infrastructure (i sent links to some foss
push servers)

or maybe rely on a third party infrastructure that we can trust enough like
Mozilla for example. ( i think both the servers and api are open source
from mozilla)

most of all this kind of function for a third party server shall be
optional so that every user is free to desactivate it

Xavier

**Comment by LuccoJ** [comment]
@xavierle As I said, you can use all the alternative push infrastructures you want, but that won't let them bypass Doze like GCM can do.

**Comment by GrayHatter** [comment]
Unless someone can give me a reason, or a direct toxcore issue, I'm very tempted to close this as either off topic, or as an Antox issue. However Antox want's to fight with Doze isn't really toxcore's concern. Did I miss something?

**Comment by LuccoJ** [comment]
The reason is in the issue's title, "Reduction of Bandwidth/Traffic Consumption", which refers to the fact that Tox itself consumes a large amount of traffic, regardless and irrespective of Doze or even Antox in particular. The amount of traffic involved has been mentioned at https://github.com/subliun/Antox/issues/6

**Comment by optimumtact** [comment]
plus any solution will almost certainly have impact on the rest of the clients

**Comment by GrayHatter** [comment]
@LuccoJ I get that, but there's no description of what would be needed in a fix. Such as what's the maximum amount of data y'all would consider acceptable from a P2P client?

**Comment by LuccoJ** [comment]
Well, I'm not the person who reported this issue so my opinion may not match theirs, but I think that at least in a mobile (or generally capped/metered) scenario, ideally the amount of data transferred should only be a function of the communication between the user and their friends, and not increase "by itself" if no user-initiated communication is taking place.

So I guess one question is whether it would be possible to have a mode that mobile clients can pick where you're not a passive part of the DHT.

Aside from that, generally speaking, mobile users usually have at least a monthly 1GB allowance, and I suppose many of them would consider it unreasonable for a "lightweight" application like IM to consume more than 10% of it, which would make around 3MB a day of idle usage an upper limit of reasonableness.

**Comment by aaannndddyyy** [comment]
I'm not a dev but a normal user.
What I expect from a lightweight messaging system is not to use more bandwidth than WhatsApp does.
So for me a reasonable boundary would be twice the usage of that.

What I don't understand with regards to those push notifications is why you would need a server at all. What is it a server can do what a normal client cannot do? Except by doing messages from different senders and send them in bulk to the recipient.
The p2p concept, and thus serverlessness is imho a key-part of tox and should not be sacrificed.
Even offline messages can be done serverless.

**Comment by aaannndddyyy** [comment]
We already have TCP relays. If they have a good uptime, they could easily maintain an idle TCP connection to mobile clients, such that the latter don't have much of battery drain. They should opt out of most activities, e.g. participation in the dht or relaying messages in groupchats - something that is necessary if you want group chats to scale but with which you do not want to burden mobile users.

**Comment by xavierle** [comment]
maybe i expressed me wrong

when i was saying servers this was to differenciate mobile light client and
any stationary stronger machine.

i agree with what said.
mobile client shall be able to disable some functions (dht and relaying
messages because of battery and bandwith)

stationary clients shall keep those functions acting as a p2p mesh of
servers not a unique central server.
in my idea every stationary client could act as a push server not a unique
central one.

to avoid necessity for push

maybe the mobile client shall check for messages time to time (define by
user) and being idle the other moment.
and when antox is being used then it is in more active connection with
other peers.

**Comment by LuccoJ** [comment]
@aaannndddyyy The issue that was being discussed isn't "push" notifications in general, but the fact in particular that, under Android 6.0, and under certain circumstances, GCM push notifications (but specifically centralized GCM ones, not just ANY push-style notification) can be the only way to wake up a device while it's "dozing", which is a new Android concept.

There is a whitelist for applications that must not "doze", which the user can enable, but it's not 100% clear to me which features get enabled again for apps after one adds them to the whitelist. https://developer.android.com/training/monitoring-device-state/doze-standby.html#support_for_other_use_cases has more information.

But GrayHatter is right that this (the "push" / GCM / Android part) should probably be discussed as an Antox-specific issue, not here at Toxcore.

**Comment by srkunze** [comment]
@GrayHatter @subliun I am uncertain where the actual problem lies and thus where to a solution should be applied and how it should look like. This was the initial reason why I first opened it on the Antox repo, then were redirected to toxcore by @Astonex.

Anyway, I am all in for closing this issue (e.g. splitting up into more tangible issues if necessary).

**Comment by LuccoJ** [comment]
@srkunze As long as it's not written off as a non-issue, though... because the problem is most definitely there for mobile clients, and it was called a toxcore issue that the Antox developer could do nothing about at https://github.com/subliun/Antox/issues/159#issuecomment-139843234

Without knowing many implementation details, I tend to think that the right solution would be for toxcore to provide a "lightweight" mode of operation, possibly one where it tends to only connect to TCP relays if at all possible, and takes part in the DHT as little as possible. Constant internet usage is simply not acceptable (often for compelling battery reasons even more than data cap reasons) on mobile clients, and mobile clients are very important for any IM service today.

**Comment by xavierle** [comment]
I agree with lucco

there  shall be a light client for mobile device (with bandwith and battery
limitations) and a full p2p client function for machines with no power or
bandwith limitations

**Comment by xavierle** [comment]
>
> I agree with lucco
>
> there  shall be a light client for mobile device (with bandwith and
battery limitations)

and a full p2p client function for machines with no power or bandwith
limitations

>
> Le 28 déc. 2015 17:47, "LuccoJ" <notifications@github.com> a écrit :
>>
>> @srkunze As long as it's not written off as a non-issue, though...
because the problem is most definitely there for mobile clients, and it was
called a toxcore issue that the Antox developer could do nothing about at
subliun/Antox#159 (comment)
>>
>> Without knowing many implementation details, I tend to think that the
right solution would be for toxcore to provide a "lightweight" mode of
operation, possibly one where it tends to only connect to TCP relays if at
all possible, and takes part in the DHT as little as possible. Constant
internet usage is simply not acceptable (often for compelling battery
reasons even more than data cap reasons) on mobile clients, and mobile
clients are very important for any IM service today.

**Comment by aaannndddyyy** [comment]
Primarily this issue is about reduction of bandwidth usage if toxcore, which would be beneficial not only for mobile clients.

If gcm is really the only method to push message notifications to a dozing phone, then its use should be optional. The user using Google play as appstore would likely use gcm by default whereas f-dried versions should do without.

But android really shouldn't rely for that feature exclusively on gcm servers...

**Comment by LuccoJ** [comment]
During the past few days, I have been using Tox on a capped data connection because I'm away from home, and surprise! My 3GB cap was completely eaten in the span of a few days. When I investigated today, it turned out that as soon as Tox starts running, I start having a bandwidth consumption between at least around 5 kilobytes per second with spikes of **more than 100*. That's... just a little excessive.
What the heck is Tox transferring? This is still the case even after I disable UDP and IPv6, which should mean I'm now only using TCP relays...

**Comment by LittleVulpix** [comment]
On my computer, tox eats about 1.5-3GB per day, with 40% download 60% upload distribution.

The traffic is about ~20-40kB/s in both directions (As in 40-80kB/s in total), with ~800-1200 open connections.

This is with UDP mode on.

I don't necessarily mind but it is a bit crazy when you think about it.

**Comment by xavierle** [comment]
to my understanding this is due to the server/relay roles of every tox
client (as there is no push central server like in xmpp each client is
working as a server :  checking which  clients are connected all the time,
and searching for messages to propagate )


this is not a problem for a stationary client ( with electrical plug and
unlimited internet)


for a mobile client this is not sustainable for battery and bandwith
consumption


i think mobile client for tox shall include  lighter client-only functions

such mobile clients would not contribute to the tox network with
server/relays functions
but i think mobile clients cannot afford to for as p2P servers/relays


i do not know if this have to be included directly in the tox core or in
the mobile app



2016-01-04 17:40 GMT+01:00 LittleVulpix <notifications@github.com>:

> On my computer, tox eats about 1.5-3GB per day, with 40% download 60%
> upload distribution.
>
> The traffic is about ~20-40kB/s in both directions (As in 40-80kB/s in
> total), with ~800-1200 open connections.
>
> This is with UDP mode on.
>
> I don't necessarily mind but it is a bit crazy when you think about it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/irungentoo/toxcore/issues/1251#issuecomment-168728046>
> .
>



--

**Comment by LuccoJ** [comment]
@xavierle with the current numbers, this is far from only a problem for mobile clients. Many people in the world have capped **landline** connections, often with caps in the order of 15 to 30GB per month. With the numbers that @LittleVulpix mentions (and which I can roughly confirm), such large caps would easily be exceeded.

Also, "every" Tox client doesn't - or at least, shouldn't - act as a relay: only those where TCP relaying has explicitly been enabled should (for potential legal reasons, as well as bandwidth usages ones, as you don't necessarily want to be relaying messages between third parties under jurisdictions where you can be held liable for them, such as Germany).

**Comment by xavierle** [comment]
at the end we come to the same idea

there shall be a difference between

:full clients" with relays function (unlimited machines)

:light clients" with client only function and minimum bandwidth
recquirements
(mostly mobile but not limited to them according tour precisions)

limitation of bandwidth consumption shall concern the light clients family
in priority

difference between full client / light client shall be an easy to
understand function.

actually i am mostly concerning by the android version.

actual


German law is crazy.... ()
i suppose in france you can claim internet service provider neutrality
protection  (translation is not good from french):
you are responsible for service only.
users are responsible for their own content.
( i am sure deutsche telekom is under such a protection. you cannot claim
for the same in germany ?)



Le 4 janv. 2016 18:00, "LuccoJ" <notifications@github.com> a écrit :
>
> @xavierle with the current numbers, this is far from only a problem for
mobile clients. Many people in the world have capped landline connections,
often with caps in the order of 15 to 30GB per month. With the numbers that
@LittleVulpix mentions (and which I can roughly confirm), such large caps
would easily be exceeded.
>
> Also, "every" Tox client doesn't - or at least, shouldn't - act as a
relay: only those where TCP relaying has explicitly been enabled should
(for potential legal reasons, as well as bandwidth usages ones, as you
don't necessarily want to be relaying messages between third parties under
jurisdictions where you can be held liable for them, such as Germany).

**Comment by LittleVulpix** [comment]
Any laws regarding "legal reasons" will not apply to tox. You cannot be responsible for any content your client may be relaying because the whole point of tox is that no-one else is able to determine what is being sent/received. They might see IP addresses and that there is traffic, but they cannot see or decrypt content. So you cannot be held responsible because no-one will know what your client is relaying :p

How scalable is tox anyway? Will this only get worse when more people join the network? Or will it get better?

**Comment by xavierle** [comment]
i have heard that germany is not so sweet for p2p usage.
in germany, there are lawsuits for p2p sharing persons because they relay
prohibited data.
the german iser may be referring to this kind of lawsuit

but it is true that tox encryption shall mask all that

**Comment by LuccoJ** [comment]
@LittleVulpix in theory, the very same thing that you said is true for RetroShare (the system that created a court precedent in Germany). But in practice, if authorities get a hold of content (such as Tox logs) from computer A that was sent from computer B, they can (from ISP logs, for instance) find out that it was relayed through computer C. User C couldn't know the content because it was encrypted, but now the authorities do know what the content was, and according to that court case, C can be held liable.

**Comment by GrayHatter** [comment]
@LuccoJ I'd like to read that court case, and the precedent?

@LittleVulpix Tox will scale to a point, I don't know where that point it, but It'll taper off at some point.

**Comment by xavierle** [comment]
@luccoj

i think that for most countries

the c person that was only a p2p relay service  provider shall be
considered as a neutral isp that is not responsible for content. (like most
telecom or internet company)

but in germany. it seems there is an exception for p2p and lawsuits were
done. against several protocols and use cases
and p2p seems not so used in germany anymore because of those lawsuits


xavier

**Comment by timofonic** [comment]
## About RetroShare:

* I tested RetroShare and has many good concepts: It brings back the good
ideas of BBS to a modern approach.
* The user experience is still difficult and a bit chaotic: They don't
release often enough with patches. I suffered some kind of MITM attack
because I was connected over his node to a chat server by a known troll
(DTS).

## Too much protocols, but all them have flaws or limitations

* I always wondered if there's efforts with other projects and have a
common protocol or some protocol-agnostic methods to reduce bandwidth
instead. This way, protocol optimization and auditing could be done by more
manpower.
* I understand competition is good, and that big projects are more prone to
corruption: Radio Free Asia's Open Technology Fund ;)

## Anonymity

* Is more important to sacrifice anonymity over certain problematic
platforms? Why we should use Tox instead other more efficient and
non-anonymous systems like Signal?


## Bandwidth

* Are there formal research about it?
* Can the protocol overhead be reduced in some way?
* What about using extremely efficient encodings for certain stuff?
** There's Codec2 for audio, it looks promising for low bandwidth networks:
I know it's not easy to integrate at this moment, I mentioned it some time
ago.

## Client homogeneity: Promote to make them support the same features

* This is very important in order to have a good user experience. Some
features aren't available on certain clients. There should be an effort in
order to avoid this situation: Testing, bounties, etc.

## Client performance

* It's very important to consider, in order to have a proper user
experience.
* Antox got unusable after some time on my Android phone. I don't know why,
but that doesn't happened on the desktop clients I tried (but my new
machine isn't so slow, too).


## My opinion abou relying on centralized services is the same, I'm
stubborn about it

* I don't like GCM, I uninstalled them on my mobile. The system performance
got a lot better, too.
* I understand Google dictatorship to be the Lords of Internet and get
forced to use their services. I'm still using them, planning to switch.
* Avoiding centralized services these days all the time is something is
practically impossible if you don't want to be some kind of neoluddite.
There's a very difficult social pressure, a network effect.
* What the reason to exist of Tox if depends in some way on centralized
services? There's Signal for encryption. Do we trust on available civil
encryption technology?

Kind regards.

**Comment by LuccoJ** [comment]
@GrayHatter you can read the complete court ruling in the original German at http://www.raschlegal.de/uploads/media/LG_Hamburg_308_O_319-12_24092012.pdf and a good article in German that covers how RetroShare works and the specifics of the court ruling is at https://irights.info/artikel/retroshare-storerhaftung-erreicht-das-„darknet/9862 while a simpler article in English about the ruling can be read at https://torrentfreak.com/anonymous-file-sharing-ruled-illegal-by-german-court-121123/

RetroShare, while offering more features than Tox (but also being quite a bit less user-friendly), is based on similar concepts, namely a DHT to discover friends' IP addresses, followed by direct communication, except in the case of relays (which was the issue here).

**Comment by LittleVulpix** [comment]
@GrayHatter that's messed up, I didn't know about that.

**Comment by GrayHatter** [comment]
@LittleVulpix what part?

**Comment by timofonic** [comment]
@LuccoJ

Why not do an interoperability effort and merge protocols progressively? I
think networks like Tox and @RetroShare MUST be perfected to correctly
scale and be resilient, despite all the possible extreme hostile
situations. Why divide resources instead merging them when this is a very
challenging task?

I would love to know the opinion of RetroShare developers about this issue
with P2P networks and possible collaboration with them. I invoke some of
the most active RetroShare developers into the discussión (I' sure I missed
many of them, I'm sorry): @csolder @defnax @electron128 @G10h4ck
@PhenomRetroShare @hunbernd @realityfabric

**Comment by G10h4ck** [comment]
@timofonic  It would be nice but I am not sure it is feasible, it seems to me that Tox overlap some RetroShare functionality but implemented in different way .

P.S. Its @csoler not csolder

**Comment by timofonic** [comment]
@G10h4ck 

Sorry, I'm not sure what happened while typing!

Can't those differences be discussed and agree on a common one? :D

**Comment by LuccoJ** [comment]
Unifying Tox with RetroShare is a little beyond the scope of this issue...

**Comment by timofonic** [comment]
@LuccoJ 

I see, sorry. I did think maybe more manpower would help to improve the issue, that's all.

**Comment by LittleVulpix** [comment]
@GrayHatter that you are liable for stuff you don't even know what is and have no way of knowing.

**Comment by GrayHatter** [comment]
Yeah, that's a scary precedent to exist.

**Comment by irungentoo** [comment]
The DHT fixes that I pushed might have improved this.

**Comment by LittleVulpix** [comment]
I am testing this right now and will post back on this tomorrow around the same time, hopefully with good news. So far it looks like the connection count decreased almost tenfold (which is almost unbeliveable).

**Comment by xavierle** [comment]
i will have a look too

**Comment by GrayHatter** [comment]
@subilun, maybe you should checkout the new toxcore for Antox?

**Comment by LittleVulpix** [comment]
@GrayHatter @irungentoo after a day of monitoring:

http://i.imgur.com/rl7HDmY.png

Previous results I don't have with me anymore, but the traffic was ~2.9GB, and 900 connections on average with peaks as high as 1800.

tl;dr - latest toxcore: 2x less data transferred and 8-10x fewer active connections (on average). Major improvement.

**Comment by xavierle** [comment]
i do not have a deep knowledge of antox and tox core.

but i maintain other technology im apps.

actually most im is organized around push notifications to get fastest
messages from server.

notifications are done by :
xmpp servers
cloud im : gcm, google cloud message ( many im apps)
websockets ( xmpp libre signal)

normal satus of the app is standby and only answering a living signal to
the nearest notification server when the main server is asking for it.
and the app really connect to receive data only after a push notification
from server.
most data transfert is done p2p without passing through server.

images files voice are already p2p even on centralised infrastructure.

maybe we can remimick this behavior for p2p with light client (in standby
mode) and full client ( with push notification relay)

tox is said to be p2p
but there is a still a kind of central introducer server.
maybe the role of this introducer have to be reevaluated.

idea 1
maybe we have to work on a get way for  light clients

continuous connection to one client if last message is less than x min with
that client

when no activity from x minutes, light client inform all contact list that
it goes in standby mode.
connections are closed and connection will wake up every y minutes by
itself (maybe 5) just to say other contact that light client is still alive
and to get any messages from them
light client never creates a connection to any client out of contact list
(no relay function)

if a client is in that standby state.
is it possible to send messages to it that will be recovered after x minutes

this will create a latency of y minutes of a conversation is inactive since
y minutes.
i think this is clearly acceptable to save resources
a p2p network was never intended to be low latency.

i think x and y of 5 minutes is very ok.
maybe this can be incremented. y = y+x
until a maximum of a minutes

with y and x starting at one minute

2
maybe on the contrary we have to include push notification function on
every full client that will act as a decentralised push servers ( maybe
like websocket)

full client are connected to many clients (full light in contact list and
out of contact list)
full clients relays push notifications to wake up the standby light clientd

light clients reduce connection to the minimum (just asnwering to live
ping) until they receive any kind of notification from any of the p2p full
client

maybe i am wrong

**Comment by GrayHatter** [comment]
@xavierle the problem with antox is that it's on android, and GCM is really the best option for battery life and actually working. Google it working hard to increase battery life, and has decided to go the software route; limiting apps and such. Rather than battery hardware R&D

**Comment by d9h02f** [comment]
I'd say they're working very hard to force everyone to become a part of the Google network or be left to die, and it seems to be working out very well. Better battery life is just a guise.

**Comment by xavierle** [comment]
my opinion is as following

there is no need to work on p2p project if it is to rely on gcm..

if i want gcm i use signal or even whatsapp or hangout..

**Comment by aaannndddyyy** [comment]
@xavierle  +1
otoh, an app draining the battery and eating all your available data volume, will not be easy to "sell"

**Comment by timofonic** [comment]
@aaannndddyyy

We can use Telegram instead. After all, you lose the most interesting
feature of ToX :)

ToX doesn't have a big foundation giving money and hidden interests. Am I
right?

Quality or quantity? :)

**Comment by GrayHatter** [comment]
Secure end to end anonymous encryption, you don't lose that from using GCM, you lose that by installing tox on any android device. If you're looking for Tox as a skype replacement, some sort of push system for your phone will be required. But if you're using tox for the same reason the concept was created. You shouldn't even be using an android.

**Comment by LuccoJ** [comment]
@xavierle +2
And others can keep saying that GCM is the "one and only" option on Android, and yet my XMPP client (Conversations) manages to work just fine, while taking up a negligible amount of battery, and without using GCM.

And XMPP is not even intrinsically a P2P protocol like Tox is, so it's pretty ironic that XMPP clients don't feel the need to go the GCM route, while Tox would give up its raison-d'être and go there.

**Comment by Astonex** [comment]
@LuccoJ The GCM requirement is for Android 6.0 only. Please read into the new doze and and standy modes introduced in that version and how it affects the ability to do networking.

**Comment by xavierle** [comment]
android 6.0 is quite virtual

most of my app users avec still with android 4
android 5 is now growing
and android 6 is nearly not existing in my communication app stats

i think users can keep with android 5 for some time to get benefits of tox

**Comment by LuccoJ** [comment]
@Astonex I have read about it, and quite extensively. Why do you assume I haven't?
However, there are several reasons why Doze and App Standby aren't necessarily such a huge problem.
There is a whitelist of apps that are exempted from App Standby (and partly from Doze) that the user can manually set. Applications exposing the right permission in the manifest can also pop up a dialog asking the user to whitelist them directly.
If they do this, they will only be accepted into the Play Store on the condition that bypassing Doze is essential for the app's "core function", but Google give an IM client that can't directly use GCM as the exact example of this in their documentation.

Please refer to http://developer.android.com/training/monitoring-device-state/doze-standby.html#whitelisting-cases for confirmation of what I said and further information.

**Comment by Astonex** [comment]
@LuccoJ I assumed you hadn't because your XMPP client working just fine without GCM didn't really relate to why Antox needs GCM (unless you happen to be on Android 6 already?).

Anyway, looking at that page you linked, it looks like Antox would be fine requesting the permissions to ignore the doze mode so it could keep on networking in the background, meaning no GCM would be needed.

Back to the main topic of this thread however, something still needs to be done about how much networking is done, and how much data is used just to run Tox, especially on a phone. Weren't offline messages supposed to have been developed last summer for GSoC? Maybe Antox could do no networking at all except checking in every now and then to look for those stored offline messages.

**Comment by subliun** [comment]
Based on the information in https://commonsware.com/blog/2015/11/11/google-anti-trust-issues.html I do not believe requesting exemption from doze through permissions is a viable solution.

**Comment by Astonex** [comment]
@subliun That post reiterates that apps must adhere to the acceptable use-cases Google has published, and that list already includes IM apps.

**Comment by subliun** [comment]
The list includes only some IM apps with some vague wording about "breaking the core function". I suppose you can try it, but you still need the not-decentralised offline messaging servers which will require more core support.

**Comment by aaannndddyyy** [comment]
Couldn't it still be a randomly chosen normal peer act as a relay like tcp relays now, and send push notifications to the mobile client?

**Comment by GrayHatter** [comment]
normal. Can't serve as Google Cloud messenger relays because they have no
way to know which and who to send to

**Comment by GrayHatter** [comment]
Correction no normal clients can't send messages via GCM because it has no
way to know which GCM ID they need send to.

There's also no way to make GCM messages distributed so we can put that
into a normal client anyways

**Comment by LuccoJ** [comment]
@Astonex The reason I mention XMPP clients is that, as you see, other people have been claiming GCM may be necessary even regardless of the Android 6.0-specific Doze / App Standby issues.

I, on the other hand, think that as long as Tox is being used in TCP relay mode, which means something like ten concurrent connections (not hundreds) and no continuous DHT requests, it should be able to perform about as well as (for instance) XMPP clients, both bandwidth-wise and battery-wise, because *it's essentially the same mechanism*.

If it currently doesn't, that must be because keepalive packets happens too often and/or communication isn't done enough in bursts when feasible. Which doesn't need GCM to be solved.
And yet, having discussed this with others, I've repeatedly encountered insistence that, somehow, this won't be good enough for Tox (even though it appears to work for things like XMPP clients).

@subliun Remember that in the worst case, Antox can always be distributed outside of Play Store, especially if Play Store have policies contrary to what a normal program needs to run without jumping through hoops. While F-Droid is an already-employed option, fortunately, Android still allows to just install standalone APKs.

**Comment by aaannndddyyy** [comment]
You misunderstood me.
I didn't talk about gcm, but about a somehow decentralized push notification system. It would still be only one, but it could be any normal, i.e. non-mobile, client taking that role. That way you do not depend on a server and do not expose too much metadata to one central authority.

**Comment by GrayHatter** [comment]
Even TCP only mode the device is still online constantly and never has a
chance to sleep so it will still kill the battery. using GCM is the best
way to preserve battery life

There is no way to preserve battery life without using some sort of push
messaging and talks does not support push messaging

**Comment by LuccoJ** [comment]
@aaannndddyyy I'm not sure I understand what people really mean when they talk about a generic "push notification solution", especially if it's decentralized. The concept of push notifications involves centralization: it is about, instead of having multiple connections to multiple entities active at the same time, keeping just one of them, which is optimized for "as much silence as possible", and only send short bursts of data when necessary.

This is not dissimilar to optimized TCP relaying in the first place.

GCM isn't magic, it's just centralization of short network communications (and special treatment by Google in Android 6.0).

**Comment by LuccoJ** [comment]
@GrayHatter You keep saying "there is no way", and I keep saying other apps, like XMPP clients, do it fine.
They are also "online constantly". With GCM, you are also "online constantly". It's still the Internet through mobile data. There is no magic.
The one difference is that GCM tries to centralize all internet-to-device communication into one connection, which it can optimize, instead of each app having its own connection.

@Astonex Do you understand why I have to keep insisting about XMPP and similar things now?

**Comment by GrayHatter** [comment]
Snow GCM isn't magic but its level of optimization greatly outstrip
anything that we can do

**Comment by LuccoJ** [comment]
Well... no.
Conversations is an XMPP client that is open source, written mainly by one person, runs on Android and doesn't eat battery at all - and yet it doesn't use GCM.
I know other programs like that.
What you say that GCM is absolutely needed for reasonable optimization, @GrayHatter simply isn't true.

**Comment by aaannndddyyy** [comment]
With gcm you are also constantly online. The difference is that you are passive until you receive a push notification. Thus the tcp connection eats almost no ressources whil you are idle.
Yes, you only keep one connection open, not hundreds. That's what the relay would be needed for. That relay would act as the server, just like tcp relays already do for peers that are not reachable via udp - all comm goes through tcp relays.

What is needed here is a passive mode. No pings every few seconds to each friend and no active participation in the dht, bwcoming active once a push was received

**Comment by GrayHatter** [comment]
I don't remember say it's absolutely needed.

But it's the easiest, most battery efficient, and soon with Android 6, the
only simple way.

Other than, there's other ways, and other than, google is centralized, or
evil, or skynet. What's wrong with GCM?

**Comment by ProMcTagonist** [comment]
@LuccoJ grayhatter isn't an antox dev

**Comment by xavierle** [comment]
hello

seems the discussion is active but still looping.

the fact about some of us does not want gcm is very simple.

there is already signal as an encrypted open source app using gcm and
centralised server.

then there are already conversations / freelab messenger for xmpp federated
client with no gcm.

tox advantage is that it does not use gcm

there is no need to be involved in a project that would be only late from
those active projects.

by definition, tox objective is to be p2p.
so this objective clearly include not to use gcm.

there might be a favultative option to activate gcm.
but by default gcm shall be full p2p.

it is clear that user will be able to whitelist apps for doze

for battery actually issue is not concerning gcm. issue is about the
keepalive connection à network activity inside the real tox core and antix
code.

toxcore is always with active conection.
this is fine for unlimited full clients.

for a limited  client  ( battery or bandwidth )
we have to use the said function of offline messages and a higher latency
delivery.
I can survive with a 5 minutes delay message check.
and when there is some messaging activity  then keeping the connection very
active until a new delay of 5 minutes for example.

there is no need to say every user if every contact is connected or not. so
there is no need for keepalive notifications and clearly not continuous
connection.
only thing that is needed is notification for delivered message.

( optionnaly saying at maximum when it was last seen but not more)
just send messages or keep them as offline messages until there are checked
by the receiver.

these are examples from recognised apps we shall learn from.
( signal/libresignal , conversations/ freelab messenger)

2

concerning server relays
if you recommend to use gcm.
there are plenty of alternative way to make push services or avoid gcm.

signal project include a push server

libresignal is a fork of signal with  a non gcm version and using
websockets instead

i prefer to run my own push server on my device always on that to use gcm.

as it seem possible to sync desktop and mobile with same account keys.
I even agree to have a desktop client running my account all time.
and that my mobile is only looking time to time what my desktop app have
received while being connected.
maybe it is a very easy way to make a kind of "relay" server.

3
it seems also that there are too independant programs.

one signaling program
one client program.

we shall merge this and signaling function to be optional for every client.






xavier

**Comment by d9h02f** [comment]
> using GCM is the best way to preserve battery life
I think it's a bit rich to claim this when GCM requires the installation of Google Play Services, battery consumer #1.

**Comment by tttom** [comment]
It seems like there is no single solution that would satisfy ever one. Although there are still arguments about battery efficiency, some are just more willing to trade efficiency for decentralization. It would be a shame if you had to run two different applications to be able to talk to everyone, and I don't see a reason why. Perhaps we should leave the wake-up mechanism to the client implementations, and restrict the discussion on toxcore architecture that can accommodate all kinds of push-services if the client is interested. Moreover, this would be required to get Tox working on non-Android devices such as iPhones.

If you want Tox to be a Skype replacement, it will have to be used by people that don't use computers nor care what push-notifications are. Without those people on board, you will be forced to use insecure communications most of the time.

**Comment by aaannndddyyy** [comment]
@xavierle 
your point 1 would require having offline messages first, which we do not have either. Also this instroduces a delay.

your point two requires multi device support, which we do not have either. 

Bith things are a good thing in their own right, however, there are users who'd only have a mobile client and no always-on desktop client at all, so I'd discard option 2.

For drastically reducing battery and bw usage, core needs some changes, but you could do it independet of offline messages and multidevice. 
An idle tcp connection does not consume much. So what is needed is a mode in which toxocre a) uses only **one** connection and b) does not participate in dht or relaying traffic for others c) does not constantly ping.

This would drastically reduce ressource consumption, and et totally do without gcm or even an alternative push server set up by Tox Foundation.

**Comment by xavierle** [comment]
thank for clarification

i was talking as functionality
i have very small knowledge of tox core code.

yours points a,  b and c were presented as possible ways in the discussion.

to express that idea i was talking about light client and full client
( same software but some core functions activated or not depending on
ressources)

i think this way is clearly good

now that we have more of r less done the easy thing : i mean talking.

maybe some great coder could do that

i feel that if points a b and c are done.
then we will have achieved something great

**Comment by GrayHatter** [comment]
So, in 24h (longer if I get distracted) I'm going to close this issue. 3 reasons. first; some of the new DHT commits will have fixed this issue (at least in part). 2nd we've clearly verred way offtrack/topic, so much I don't thin. We're coming back. And 3rd, I don't think anyone wants GCM code in toxcore.

If you still think we need a toxcore solution to this antox problem. Open a new issue here, describe the problem really well, and describe what's required for a fix.

**Comment by aaannndddyyy** [comment]
@GrayHatter: This ticket is not about gcm - gcm was one proposed 
solution.

And the issue is not an Antox-only issue, it is an issue with any mobile 
client, be it on Android or iOS. Even desktop user with really low data 
caps would benefit.

What would be needed for a fix was already outlined several times:
a mode in which a client maintains only **one** tcp connection and no 
udp connections at all, does no forwarding (I think current tcp-only 
mode already does that part) and pings or keepalives at most every ten 
minutes.

**Comment by xavierle** [comment]
thank you for sum up andy

this seems to me to be linked to tow core
so that it is easy to implement for every client later

**Comment by GrayHatter** [comment]
@aaannndddyyy this would be perfect time for you to open an issue to that effect. This thread has become unmanageable as an issue that can be marked as solved to everyone. 

Not to mention the OP and the title mentions neither of those suggestions.

**Comment by GrayHatter** [comment]
My goal is to make issues easy to work on any close. This thread would take longer to read, than to fix at this point, so it's no longer a code issue, but a discussion on what color to paint the bikeshed.

**Comment by xavierle** [comment]
this is clearly a discussion
but at some points a way have been shown

no problem to close this an recreate a more defined one.

why do you seem borred/ angry about this kind of discussion about the
project ?

would you prefer noone is interest in the project to avoid such debate ?

personally i do not enjoy the way you say thay you will close the issue
(even if i agree the reason)
neither your sarcasm about discussion and painting.

hopefully in the future issue your smile would be back on track

**Comment by aaannndddyyy** [comment]
@GrayHatter If you think that my description is at least somewhat 
reasonable and viable, then I will create a new issue with that content 
tonight. Right now am kinda busy.
If you think it's outright bs or unfeasable, then no need to create a 
ticket that will immediatelly get closed anyhow.

Feedback appreciated.

**Comment by ameenross** [comment]
This seems to be a meta issue. There's probably a wide range of possibilities to reduce bandwidth. I think each should be discussed separately as well.

Also, FWIW, I'm getting vastly better battery consumption with Android 6 (cm-13) compared to Android 5.1 (cm-12.1). And I haven't used Google apps for a few years (not installed).

**Comment by GrayHatter** [comment]
@aaannndddyyy I think your description is very workable.

I don't have a problem with this issue, nor with it turning into a
discussion. I'm not going to lock it either. Only going to close it, simply
because the title and OP are actually completed. So now this is simply a
discussion about how best to deal with android M and that's not an issue
for toxcore.

**Comment by srkunze** [comment]
@aaannndddyyy @GrayHatter Good suggestion. I am going to close this issue as soon as I see @aaannndddyyy's proposal.

This issue led to a number of possible ways of fixing things. So, now it's time to seize discussion and go coding on specific proposals.

**Comment by LuccoJ** [comment]
@GrayHatter except that toxcore **still** uses a lot more bandwidth (and hence, on mobile devices, power) than other typical IM systems, even in TCP relaying mode, where it really has no reason to.
I like what @aaannndddyyy is proposing (fewer connections, fewer keepalives), so I'm all for going ahead with it, but why does this issue need to be closed, given all that has only been stated but *not* done yet?
The bandwidth/traffic consumption issue is all still there, for now.

**Comment by xavierle** [comment]
i was a bit surprised but reasonable reasons were pointed out

this thread became  a discussion and not an issue.
as said the text is longer than the issue

it will be closed as an issue but not locked.
discussion will remain open.

as a results of the discussion, more precisely  defined issues will have to
be created.

andy made a good sum up of the possible ways.

**Comment by aaannndddyyy** [comment]
@all FYI: #1501

**Comment by srkunze** [comment]
Discussion can be continued of course. :)

**Comment by aaannndddyyy** [comment]
is it better to discuss here ? 
If we discuss in the new one, it might become long again too and then become unreadable again.

**Comment by GrayHatter** [comment]
It's better to discuss either on the ML or IRC. Here is fine, but don't let
any of the other issues get sidetracked because then we'll have the same
problem.

Issues only become too long when other issues are raised and discussed.

**Comment by aaannndddyyy** [comment]
Could someone who has whatsapp and/or signal please profile that app for a week or so, as to see how much bandwidth those use when idle or with only text messages?

Would be nice to have a value to compare with.

**Comment by xavierle** [comment]
for january

whatsapp 179ko text only
freelab messenger/ conversations : 5mo with some images
signal : 3mo with some images

aml this recquires very few data

**Comment by aaannndddyyy** [comment]
Thank you.
This is SIGNIFICANTLY less than tox uses.
Ideally, tox on mobile should have a traffic consumption of that order...

**Comment by srkunze** [comment]
@GrayHatter That's indeed a very good benchmark test.

@aaannndddyyy Is there someway reproducibly testing this?

**Comment by xavierle** [comment]
i do not claim any accuracy on the information i sent.
i took the information from android bandwith app usage

i do not know if i use 3 apps the same way.
just all 3 are all day active and daily used with 5 persones each

**Comment by mray** [comment]
I feel foolish to have encouraged others to try Antox. It drained the bandwidth of a whole month within days. this should at least be obvious before using.

**Comment by LittleVulpix** [comment]
Whenever someone asks about a mobile client, I have to:

a) explain there are no accounts and so they will need to create a new profile and add me again
b) explain it will kill their battery if it is running
c) explain it will eat most of their bandwidth so it is to be used only when connected via WIFI or if they have an unlimited data plan
d) explain that there are no offline messages

Not really antox's fault, but I feel that antox is like trying to make a portable construction for your desktop. Yeah you can make it but you'll need heavy batteries and your back will probably hurt after 15 minutes.

I agree with @mray - users should be clearly told that antox will do all of the above to their device/data plan.

**Comment by ProMcTagonist** [comment]
Alternatively -- I'm simply not able to evangelize Antox until those things are fixed
